### PR TITLE
optimize bias accumulation in avx backend

### PR DESCRIPTION
**Pull request by beru**
This PR gives about 5 % speed-up in `example_cifar_train`.

### CODE REVIEW

**Comment by bhack** [comment]
@beru Are you working with Intel cpus? We could investigate to add another engine for [MKL-DNN](https://github.com/01org/mkl-dnn).

**Comment by beru** [comment]
@bhack You may do so as much as you want. It should be a lot faster than current avx backend.

**Comment by bhack** [comment]
I've open an issue to see if there is also Atom in the roadmap https://github.com/01org/mkl-dnn/issues/12

File: tiny_dnn/core/kernels/conv2d_grad_op_avx.h [hunk 1]
```diff
@@ -439,8 +485,9 @@ void avx_conv2d_5x5_back_kernel_one(const core::conv_params& params,
 
     // accumulate db
     if (params.has_bias) {
-        //fvec_t& db = *in_grad[2];
-        
+#if 1
```
Comment (beru) [inline tiny_dnn/core/kernels/conv2d_grad_op_avx.h]: dead code remains 4ever.

Comment (nyanp) [inline tiny_dnn/core/kernels/conv2d_grad_op_avx.h]: I prefer to use comments & github history. ```#if 1 - #else - #endif``` block is unclear. Do we need to leave this else block?

Comment (beru) [inline tiny_dnn/core/kernels/conv2d_grad_op_avx.h]: No, we don't. I should have kept reference code in my brain.

File: tiny_dnn/core/kernels/conv2d_grad_op_avx.h [hunk 2]
```diff
@@ -41,6 +41,52 @@ namespace kernels {
...
 // float ver
 template <typename Allocator>
+inline void accumulate_db(const index3d<serial_size_t>&        out,
+                          const std::vector<float, Allocator>& curr_delta,
+                          std::vector<float, Allocator>&       db) {
+    if (out.width_ == 1 && out.height_ == 1) {
+        size_t nblocks = out.depth_ / 8;
+        size_t remainder = out.depth_ & 7;
+        for (size_t i = 0; i < nblocks; ++i) {
+            _mm256_storeu_ps(&db[i*8], _mm256_add_ps(_mm256_loadu_ps(&db[i*8]), _mm256_loadu_ps(&curr_delta[i*8])));
```
Comment (beru) [inline tiny_dnn/core/kernels/conv2d_grad_op_avx.h]: If you impose `spaces < 80` rule, please add it to [Wiki](https://github.com/tiny-dnn/tiny-dnn/wiki/Implementation-Guides#coding-rules) and [this document](https://github.com/tiny-dnn/tiny-dnn/blob/master/docs/developer_guides/How-to-contribute.md)

_Expect heavy changes!_

**Comment by beru** [comment]
@edgarriba @nyanp I updated the code so please review it again.

**Comment by edgarriba** [comment]
@beru OMG, too heavy, no? 😅

**Comment by beru** [comment]
@edgarriba Hmmm, yes. I'll revert it then. :racehorse:

File: tiny_dnn/core/kernels/conv2d_grad_op_avx.h (hunk 2 above)
Comment (bhack) [inline tiny_dnn/core/kernels/conv2d_grad_op_avx.h]: We are waiting for https://github.com/tiny-dnn/tiny-dnn/pull/363

File: tiny_dnn/core/kernels/conv2d_grad_op_avx.h [hunk 3]
```diff
@@ -41,13 +19,71 @@ namespace kernels {
...
+            0, 0, 0, 0,
+            0, 0, 0, 0,
+        };
+        __m256i mask = _mm256_loadu_si256((const __m256i*)
+                                          (masks + 8 - remainder));
+        for (size_t outc = 0; outc < out.depth_; outc++) {
+            const float *delta = &curr_delta[out.get_index(
+                                                0,
+                                                0,
+                                                (serial_size_t)outc
```
Comment (edgarriba) [inline tiny_dnn/core/kernels/conv2d_grad_op_avx.h]: Could you use type safe cast here? `static_cast<size_t>`

File: tiny_dnn/core/kernels/conv2d_grad_op_avx.h [hunk 4]
```diff
@@ -56,16 +92,25 @@ void avx_conv2d_5x5_back_kernel_one(const core::conv_params& params,
...
-    static const __m256 mask = _mm256_castsi256_ps(_mm256_setr_epi32(-1, -1, -1, -1, -1, 0, 0, 0));
+    static const __m256i imask = _mm256_setr_epi32(
+        -1, -1, -1, -1, -1, 0, 0, 0);
+    static const __m256 mask = _mm256_castsi256_ps(
+        _mm256_setr_epi32(-1, -1, -1, -1, -1, 0, 0, 0));
     // propagate delta to previous layer
     if (w_stride == 1 && out.width_ >= 4) {
         const serial_size_t nblocks = out.width_ / 4;
-        for (serial_size_t inc = 0; inc < in.depth_; ++inc, pdelta_dst_org += in_padded_area) {
+        for (serial_size_t inc = 0;
```
Comment (edgarriba) [inline tiny_dnn/core/kernels/conv2d_grad_op_avx.h]: `size_t` ?

Comment (beru) [inline tiny_dnn/core/kernels/conv2d_grad_op_avx.h]: Maybe not yet, the type of `tiny_dnn::conv_params::Params::in' is `index3d<serial_size_t>` at the moment.

Comment (edgarriba) [inline tiny_dnn/core/kernels/conv2d_grad_op_avx.h]: Right. Another big todoo in the list

File: tiny_dnn/core/kernels/conv2d_grad_op_avx.h (hunk 3 above)
Comment (beru) [inline tiny_dnn/core/kernels/conv2d_grad_op_avx.h]: I did change the code. But there're still C-style casts.

**Comment by nyanp** [comment]
@beru 
LGTM 👍

**Pull request merged**